graphql/query_resolver.go

```diff
@@ -1130,6 +1130,102 @@ func (r *queryResolver) Waterfall(ctx context.Context, options WaterfallOptions)
 	return results, nil
 }
 
+// TaskHistory is the resolver for the taskHistory field.
+func (r *queryResolver) TaskHistory(ctx context.Context, options TaskHistoryOpts) (*TaskHistory, error) {
+	projectId, err := model.GetIdForProject(ctx, options.ProjectIdentifier)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("fetching project '%s': %s", options.ProjectIdentifier, err.Error()))
+	}
+
+	if options.CursorParams == nil {
+		return nil, InputValidationError.Send(ctx, "must specify cursor params")
+	}
+
+	taskID := options.CursorParams.CursorID
+	includeCursor := options.CursorParams.IncludeCursor
+
+	foundTask, err := task.FindOneId(ctx, taskID)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("fetching task '%s': %s", taskID, err.Error()))
+	}
+	if foundTask == nil {
+		return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("task '%s' not found", taskID))
+	}
+	taskOrder := foundTask.RevisionOrderNumber
+
+	opts := model.FindTaskHistoryOptions{
+		TaskName:     options.TaskName,
+		BuildVariant: options.BuildVariant,
+		ProjectId:    projectId,
+		Limit:        options.Limit,
+	}
+
+	if options.CursorParams.Direction == TaskHistoryDirectionBefore {
+		if includeCursor {
+			opts.UpperBound = utility.ToIntPtr(taskOrder)
+		} else {
+			opts.UpperBound = utility.ToIntPtr(taskOrder - 1)
+		}
+	}
+	if options.CursorParams.Direction == TaskHistoryDirectionAfter {
+		if includeCursor {
+			opts.LowerBound = utility.ToIntPtr(taskOrder)
+		} else {
+			opts.LowerBound = utility.ToIntPtr(taskOrder + 1)
+		}
+	}
+
+	activeTasks, err := model.FindActiveTasksForHistory(ctx, opts)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("finding active tasks: %s", err.Error()))
+	}
+
+	if len(activeTasks) > 0 && opts.UpperBound == nil {
```

Comment:
I think it would be helpful for the future to put some comments to explain why you're using this if statements to change the bounds for inactive tasks.

Reply:
And/or could make a separate function in model that just gets all tasks (so you can also move the sort logic below there and etc), as well as the new logic you're adding in DEVPROD-16060, and then FindActive and FindInactive can be private functions (just an idea though)

Reply:
Refactored, let me know what you think now!

Reply:
refactor and comments are 🔥